<!doctype html>
<html class="no-js" lang="en">

<head>
  <!-- Global site tag (gtag.js) - Google Analytics -->
  <script async src="https://www.googletagmanager.com/gtag/js?id=UA-134931760-1"></script>
  <script>
    window.dataLayer = window.dataLayer || [];
    function gtag(){dataLayer.push(arguments);}
    gtag('js', new Date());

    gtag('config', 'UA-134931760-1');
  </script>

  <meta charset="utf-8">
  <meta http-equiv="x-ua-compatible" content="ie=edge">
  <title>try ... catch VS if ... else. What, when and why?</title>
  <link rel="icon" type="image/x-icon" href="/favicon.ico" />
  <meta name="description" content="This article is designed for two types of developers: 
 - who is not familiar with exceptions 
 - who tries to find more reasonable use of exceptions ...">
  <meta name="viewport" content="width=device-width, initial-scale=1, shrink-to-fit=no">

  <link rel="stylesheet" href="../../css/main.css">

  <script src="https://cdnjs.cloudflare.com/ajax/libs/jquery/3.3.1/jquery.min.js"></script>
  <script>window.jQuery || document.write('<script src="../../js/vendors/jquery-3.3.1.min.js"><\/script>')</script>

  <script async src="//pagead2.googlesyndication.com/pagead/js/adsbygoogle.js"></script>
  <script>
       (adsbygoogle = window.adsbygoogle || []).push({
            google_ad_client: "ca-pub-6974184241884155",
            enable_page_level_ads: true
       });
  </script>
</head>

<body>
  <!--[if lte IE 9]>
    <p class="browserupgrade">You are using an <strong>outdated</strong> browser. Please <a href="https://browsehappy.com/">upgrade your browser</a> to improve your experience and security.</p>
  <![endif]-->
  <header class="page-header js-page-header">
    <a class="page-header-logo-container" href="https://weekly-geekly.github.io/index.html"></a>
    <div class="page-header-text">Geekly Articles each Day</div>
  </header>
  <nav class="page-headings-container js-page-headings-container"></nav>
  <div class="tools-bar js-tools-bar">
    <!-- <a href="../../search.html" title="Search">üîé</a> -->
    <a class="js-list-of-headings-button" data-state="closed" href="#" title="Headings">üìú</a>
    <a class="js-go-to-top-button" href="#" title="Go to Top">‚¨ÜÔ∏è</a>
    <a class="js-go-to-bottom-button" href="#" title="Go to Bottom">‚¨áÔ∏è</a>
  </div>
  <a href="http://bit.ly/donateToWeeklyGeekly" class="donate-btn">DONATE</a>
  <section class="page js-page"><h1>try ... catch VS if ... else. What, when and why?</h1><div class="post__text post__text-html js-mediator-article">  This article is designed for two types of developers: <br>  - who is not familiar with exceptions <br>  - who tries to find more reasonable use of exceptions <br><br>  In the article I will talk about the very basics of exceptions, about how you can do without them, as well as about how I see the correct use of certain features of the language ... <br><br>  The evolution of programming languages ‚Äã‚Äãsometimes leads to dramatic changes in the developers' worldview.  In the world of PHP, this happened when the fifth branch appeared, which brought a new object model, new sets of built-in functions and new error handling methods ... 
      <br>
        <script async src="//pagead2.googlesyndication.com/pagead/js/adsbygoogle.js"></script>
        <ins class="adsbygoogle"
          style="display:block; text-align:center;"
          data-ad-layout="in-article"
          data-ad-format="fluid"
          data-ad-client="ca-pub-6974184241884155"
          data-ad-slot="8945601208"></ins>
        <script>
          (adsbygoogle = window.adsbygoogle || []).push({});
        </script>
      <br>
      Those who started to get acquainted with PHP (hereinafter I will imply the fifth version), after another procedural programming language, did not understand what was happening in the transition from 4k to 5k and continued to draw code with familiar functions that are on the same level relation to each other, as well as each action is checked for a successful return code.  But those who knew about exceptions and classes ... <br><a name="habracut"></a><br><br><h3>  Theory </h3><br>  Exception - a signal sent by the program to the interpreter, about the occurrence of an abnormal (exceptional) situation during the execution of a code. <br><br>  In PHP, working with exceptions is based on the same principles as everywhere else.  To generate an exception, use the throw operator, to which an exception object is passed, which takes two optional parameters in the constructor: an exception message and an exception code. <br><br><blockquote><code>throw new Exception(\\\"This is exception message\\\", $exception_code); <br></code> </blockquote><br><br>  To catch an exception, use the try ... catch construction.  In the try block, operations are performed that can lead to an exception, and the catch block allows you to decide what to do if an exception has been thrown. <br><blockquote> <code>try { <br> throw new Exception(\\\"Exception message\\\"); <br> echo \\\"That code will never been executed\\\"; <br> } catch (Exception $e) { <br> echo $e-&gt;getMessage(); // \\\"Exception message\\\" <br> } <br></code> </blockquote><br><br>  As you can see from the example, when an exception is thrown, the rest of the code in the try block will not be executed, and control will be passed to the catch statement, in which we specify what the object to which the thrown exception will be passed will be called (in our case, $ e) .  Inside the block of the catch operator, based on the data from the exception, we can apply some action depending on the situation.  Now we just brought out the message that was transmitted by the exception. <br><br>  The Exception object (which, by the way, can be inherited) has the following set of final (final) methods: <br><blockquote> <code>final function getMessage(); //   <br> final function getCode(); //   <br> final function getFile(); //      <br> final function getLine(); //   <br> final function getTrace(); //    <br> final function getTraceAsString(); //       <br></code> </blockquote><br><br><h3>  Practice </h3><br>  When is it convenient to use exceptions?  Yes, always, when a function or method can come to an error situation!  Using the built-in functions mysql_connect () and mysql_select_db () as an example, if they threw exceptions during a connection error and database selection, respectively, the code would look like this: <br> <code>try { <br> mysql_connect($hostname, $username, $password); <br> mysql_select_db($dbname); <br> } catch (Exception $e) { <br> echo $e-&gt;getMessage(); //     ,     <br> } <br></code> <br><br>  And depending on the result, we would: <br><br>  * would successfully join the DBMS and select the database (the code in the catch block would not be executed) <br>  * if the connection with the DBMS failed, they would output the corresponding error and rolled up the session <br>  * in case of a database selection error, the user would be informed about this trouble. <br><br>  Now it's time to ask: \\\ "And why do we need such a complicated thing, if you can use the if operator? \\\". <br><br>  The first answer is: \\\ "Because the code is more readable, and error handling is displayed in a separate \\\ block."  If we used the if statement, the code would look like this: <br><blockquote> <code>$connId = mysql_connect($hostname, $username, $password); <br> if (false == $connId) { <br> echo \\\"   \\\"; <br> } <br> <br> $flag = mysql_select_db($dbname); <br> if (false == $flag) { <br> echo \\\"   .\\\"; <br> } <br></code> </blockquote><br><br>  Agree that in the previous version the code looks more understandable.  But that is not all.  Now it‚Äôs not necessary to deal with these errors directly to the developer - it‚Äôs enough to throw an exception, but you can handle it at a higher level.  You can also pass exceptions with a chain (chain) to the top: <br><blockquote> <code>class MyException extends Exception {} <br> <br> try { <br> try { <br> //... <br> throw new Exception(\\\"inner\\\"); <br> //... <br> } catch (Exception $e) { <br> throw new MyException(\\\"outer\\\"); <br> } <br> } catch (MyException $e) { <br> echo $e-&gt;getMessage(); // \\\"outer\\\" <br> } <br></code> </blockquote><br><br><h3>  Performance </h3><br>  Trying to reach the truth, he conducted several experiments with various types of functions. <br>  <em>The first type</em> returned the status true and was checked by the operators if ... else <br>  <em>The second type</em> returned the status false and was checked by the if ... else statements <br>  <em>The third type</em> simply performed the actions and returned nothing.  Tested with try ... catch <br>  <em>The fourth type</em> always threw an exception and checked it in a try ... catch <br><br>  Results: <br><blockquote>  True: 0.72382092475891 <br>  False: 0.85190796852112 <br>  No exception: 0.72565317153931 <br>  Exception: 14.176206827164 </blockquote><br><br>  As expected - throwing exceptions is a rather expensive operation, but both options for successful execution passed level tests. <br>  Tests were conducted on primitive addition functions, but in 1cc iterations <br><br><h3>  findings </h3><br>  <strong>When to use exceptions?</strong>  Always, when an error or non-standard behavior of the program is implied, as well as when the decision to process the result has to be shifted to a higher level. <br>  <strong>And what to do with the if operator and boolean statuses of working off functions?</strong>  Leave them.  But only where they are really needed.  Ie where the logical operator implies the use of the result in calculations rather than controlling the flow of execution.  Ie all successful completion of functions no longer need to notify operator return true, if this logical value is not useful for further calculations.  And at the same time, all statuses of completion of functions with errors can be changed from the return false format to the throw format new Exception () <br>  <strong>What you need to remember?</strong>  The use of exceptions assumes that all code is executed with the status true (without errors), but if an error occurs, then there is always room for processing it in the catch block. </div><p>Source: <a href="https://habr.com/ru/post/21376/">https://habr.com/ru/post/21376/</a></p>
<section class="more-articles-navigation-panel js-more-articles-navigation-panel">
<h4>More articles:</h4>
<nav class="list-of-articles-container js-list-of-articles-container"><ul class="list-of-pages js-list-of-pages">
<li><a href="../213749/index.html">Homemade phase laser range finder</a></li>
<li><a href="../213751/index.html">Ten most controversial theses about time</a></li>
<li><a href="../213755/index.html">MWC 2014. App Planet. Intel Software</a></li>
<li><a href="../213757/index.html">Thinking out loud: TV firewall for older people</a></li>
<li><a href="../213759/index.html">Samsung introduced the Gear Fit fitness bracelet at # MWC2014</a></li>
<li><a href="../213761/index.html">Report from the Integrated Systems Europe-2014</a></li>
<li><a href="../213763/index.html">Exchanger MtGox stopped working</a></li>
<li><a href="../213765/index.html">Elastic Search in enterprise projects</a></li>
<li><a href="../213767/index.html">The first flash card microSD 128 GB</a></li>
<li><a href="../213769/index.html">10 trends in the mobile gaming market in 2014</a></li>
</ul></nav>
</section><br />
<a href="../../allArticles.html"><strong>All Articles</strong></a>
<script src="../../js/main.js"></script>

<!-- Yandex.Metrika counter -->
<script type="text/javascript" >
  (function (d, w, c) {
      (w[c] = w[c] || []).push(function() {
          try {
              w.yaCounter52496797 = new Ya.Metrika({
                  id:52496797,
                  clickmap:true,
                  trackLinks:true,
                  accurateTrackBounce:true,
                  webvisor:true
              });
          } catch(e) { }
      });

      var n = d.getElementsByTagName("script")[0],
          s = d.createElement("script"),
          f = function () { n.parentNode.insertBefore(s, n); };
      s.type = "text/javascript";
      s.async = true;
      s.src = "https://mc.yandex.ru/metrika/watch.js";

      if (w.opera == "[object Opera]") {
          d.addEventListener("DOMContentLoaded", f, false);
      } else { f(); }
  })(document, window, "yandex_metrika_callbacks");
</script>
<noscript><div><img src="https://mc.yandex.ru/watch/52496797" style="position:absolute; left:-9999px;" alt="" /></div></noscript>

<!-- Google Analytics -->
  <script>
    window.ga = function () { ga.q.push(arguments) }; ga.q = []; ga.l = +new Date;
    ga('create', 'UA-134931760-1', 'auto'); ga('send', 'pageview')
  </script>
  <script src="https://www.google-analytics.com/analytics.js" async defer></script>

</section>

  <footer class="page-footer">
    <div class="page-footer-legal-info-container page-footer-element">
      <p>
        Weekly-Geekly | <span class="page-footer-legal-info-year js-page-footer-legal-info-year">2019</span>
      </p>
    </div>
    <div class="page-footer-counters-container page-footer-element">
      <a class="page-footer-counter-clustrmap" href='#'  title='Visit tracker'><img src='https://clustrmaps.com/map_v2.png?cl=698e5a&w=271&t=t&d=6iCFw7uJz0zcOaoxz5k5PcLCJUzv2WG8G5V8M3U6Rc4&co=3a3a3a&ct=ffffff'/></a>
    </div>
  </footer>
</body>

</html>